// @ilovepdf/ilovepdf-js/samples/sign_advanced.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const __1 = __importDefault(require(".."));
const dotenv_1 = __importDefault(require("dotenv"));
const SignatureFile_1 = __importDefault(require("@ilovepdf/ilovepdf-js-core/tasks/sign/elements/SignatureFile"));
const Signer_1 = __importDefault(require("@ilovepdf/ilovepdf-js-core/tasks/sign/receivers/Signer"));
const Validator_1 = __importDefault(require("@ilovepdf/ilovepdf-js-core/tasks/sign/receivers/Validator"));
const Witness_1 = __importDefault(require("@ilovepdf/ilovepdf-js-core/tasks/sign/receivers/Witness"));
// Load env vars.
dotenv_1.default.config();
const instance = new __1.default(process.env.PUBLIC_KEY);
const task = instance.newTask('sign');
task.start()
    .then(() => {
    return task.addFile('<FILE_URL>');
})
    .then(file => {
    // Multiple elements.
    const signatureFile = new SignatureFile_1.default(file, [
        {
            type: 'signature',
            position: '300 -100',
            pages: '1',
            size: 28,
            color: '#000000',
            font: 'Arial-Unicode-MS',
        },
        {
            type: 'initials',
            position: '300 -100',
            pages: '1',
            size: 28,
            color: '#000000',
        },
        {
            type: 'text',
            position: '300 -100',
            pages: '1',
            size: 28,
            color: '#000000',
            content: 'This is a text field.',
        },
    ]);
    return signatureFile;
})
    .then(signatureFile => {
    // RECEIVERS.
    const signer = new Signer_1.default('Signer', 'signer@email.com');
    // Add elements to the receivers that need it.
    signer.addFile(signatureFile);
    const validator = new Validator_1.default('Validator', 'validator@email.com');
    const witness = new Witness_1.default('Witness', 'witness@email.com');
    // Add all receivers to the Sign task.
    task.addReceiver(signer);
    task.addReceiver(validator);
    task.addReceiver(witness);
})
    .then(() => __awaiter(void 0, void 0, void 0, function* () {
    const brandLogo = yield task.addFile('<LOGO_URL>');
    return task.process({
        brand_name: 'My company name',
        brand_logo: brandLogo.serverFilename,
        reminders: true,
        uuid_visible: true,
    });
}))
    .then(response => {
    console.log(response);
});
