// @ilovepdf/ilovepdf-js/ILovePDFApi.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const JWT_1 = __importDefault(require("@ilovepdf/ilovepdf-js-core/auth/JWT"));
const TaskFactory_1 = __importDefault(require("@ilovepdf/ilovepdf-js-core/tasks/TaskFactory"));
const ILovePDFCoreApi_1 = __importDefault(require("@ilovepdf/ilovepdf-js-core/ILovePDFCoreApi"));
const XHRPromise_1 = __importDefault(require("./XHRPromise"));
class ILovePDFApi {
    constructor(publicKey, params = {}) {
        this.xhr = new XHRPromise_1.default;
        this.auth = new JWT_1.default(this.xhr, publicKey, undefined, params);
        this.taskFactory = new TaskFactory_1.default();
    }
    /**
     * @inheritdoc
     */
    newTask(taskType) {
        return this.taskFactory.newTask(taskType, this.auth, this.xhr);
    }
    /**
     * @inheritdoc
     */
    getSignatureStatus(signatureToken) {
        return __awaiter(this, void 0, void 0, function* () {
            return ILovePDFCoreApi_1.default.getSignatureStatus(this.auth, this.xhr, signatureToken);
        });
    }
    /**
     * @inheritdoc
     */
    getSignatureList(page = 0, pageLimit = 20, filters = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            return ILovePDFCoreApi_1.default.getSignatureList(this.auth, this.xhr, page, pageLimit, filters);
        });
    }
    /**
     * @inheritdoc
     */
    voidSignature(signatureToken) {
        return __awaiter(this, void 0, void 0, function* () {
            return ILovePDFCoreApi_1.default.voidSignature(this.auth, this.xhr, signatureToken);
        });
    }
    /**
     * @inheritdoc
     */
    increaseSignatureExpirationDays(signatureToken, daysAmount) {
        return __awaiter(this, void 0, void 0, function* () {
            return ILovePDFCoreApi_1.default.increaseSignatureExpirationDays(this.auth, this.xhr, signatureToken, daysAmount);
        });
    }
    /**
     * @inheritdoc
     */
    sendReminders(signatureToken) {
        return __awaiter(this, void 0, void 0, function* () {
            return ILovePDFCoreApi_1.default.sendReminders(this.auth, this.xhr, signatureToken);
        });
    }
    /**
     * @inheritdoc
     */
    downloadOriginalFiles(signatureToken) {
        return __awaiter(this, void 0, void 0, function* () {
            return ILovePDFCoreApi_1.default.downloadOriginalFiles(this.auth, this.xhr, signatureToken);
        });
    }
    /**
     * @inheritdoc
     */
    downloadSignedFiles(signatureToken) {
        return __awaiter(this, void 0, void 0, function* () {
            return ILovePDFCoreApi_1.default.downloadSignedFiles(this.auth, this.xhr, signatureToken);
        });
    }
    /**
     * @inheritdoc
     */
    downloadAuditFiles(signatureToken) {
        return __awaiter(this, void 0, void 0, function* () {
            return ILovePDFCoreApi_1.default.downloadAuditFiles(this.auth, this.xhr, signatureToken);
        });
    }
    /**
     * @inheritdoc
     */
    getReceiverInfo(receiverTokenRequester) {
        return __awaiter(this, void 0, void 0, function* () {
            return ILovePDFCoreApi_1.default.getReceiverInfo(this.auth, this.xhr, receiverTokenRequester);
        });
    }
    /**
     * @inheritdoc
     */
    fixReceiverEmail(receiverTokenRequester, email) {
        return __awaiter(this, void 0, void 0, function* () {
            return ILovePDFCoreApi_1.default.fixReceiverEmail(this.auth, this.xhr, receiverTokenRequester, email);
        });
    }
    /**
     * @inheritdoc
     */
    fixReceiverPhone(receiverTokenRequester, phone) {
        return __awaiter(this, void 0, void 0, function* () {
            return ILovePDFCoreApi_1.default.fixReceiverPhone(this.auth, this.xhr, receiverTokenRequester, phone);
        });
    }
}
exports.default = ILovePDFApi;
